File: src/api/performance.ts
import { API_BASE_URL, createAuthHeaders, handleApiError } from './auth';

// Enums
export enum PerformanceRating {
	OUTSTANDING = 'OUTSTANDING',
	EXCEEDS_EXPECTATIONS = 'EXCEEDS_EXPECTATIONS',
	MEETS_EXPECTATIONS = 'MEETS_EXPECTATIONS',
	NEEDS_IMPROVEMENT = 'NEEDS_IMPROVEMENT',
	UNSATISFACTORY = 'UNSATISFACTORY',
}

// Types
export interface PerformanceRequest {
	periodStart: string; // ISO date
	periodEnd: string; // ISO date
	newClientsAcquired: number;
	policiesRenewed: number;
	totalPremiumCollected: number;
	salesTarget: number;
	salesAchieved: number;
	clientRetentionRate?: number;
	customerSatisfactionScore?: number;
	attendanceScore: number;
	qualityScore: number;
	managerFeedback?: string;
}

export interface PerformanceResponse {
	id: number;

	// Agent information
	agentId: number;
	agentFirstName: string;
	agentLastName: string;
	agentEmail: string;

	// Manager information
	managerId?: number;
	managerFirstName?: string;
	managerLastName?: string;
	managerEmail?: string;

	// Performance metrics
	periodStart: string;
	periodEnd: string;
	newClientsAcquired: number;
	policiesRenewed: number;
	totalPremiumCollected: number;
	salesTarget: number;
	salesAchieved: number;
	achievementPercentage: number;
	clientRetentionRate?: number;
	customerSatisfactionScore?: number;
	rating: PerformanceRating;
	managerFeedback?: string;
	attendanceScore: number;
	qualityScore: number;
	overallScore: number;

	// Timestamps
	createdAt: string;
	updatedAt: string;
}

// Helper functions
export const calculatePerformanceMetrics = (
	data: PerformanceRequest
): {
	achievementPercentage: number;
	overallScore: number;
	rating: PerformanceRating;
} => {
	// Calculate achievement percentage
	const achievementPercentage = (data.salesAchieved / data.salesTarget) * 100;

	// Calculate overall score (weighted average)
	const salesWeight = 0.4;
	const retentionWeight = 0.2;
	const satisfactionWeight = 0.2;
	const attendanceWeight = 0.1;
	const qualityWeight = 0.1;

	const salesScore = achievementPercentage;
	const retentionScore = data.clientRetentionRate || 0;
	const satisfactionScore = data.customerSatisfactionScore || 0;
	const attendanceScoreValue = data.attendanceScore;
	const qualityScoreValue = data.qualityScore;

	const overallScore =
		salesScore * salesWeight +
		retentionScore * retentionWeight +
		satisfactionScore * satisfactionWeight +
		attendanceScoreValue * attendanceWeight +
		qualityScoreValue * qualityWeight;

	// Determine rating based on overall score
	let rating: PerformanceRating;
	if (overallScore >= 90) rating = PerformanceRating.OUTSTANDING;
	else if (overallScore >= 80) rating = PerformanceRating.EXCEEDS_EXPECTATIONS;
	else if (overallScore >= 70) rating = PerformanceRating.MEETS_EXPECTATIONS;
	else if (overallScore >= 60) rating = PerformanceRating.NEEDS_IMPROVEMENT;
	else rating = PerformanceRating.UNSATISFACTORY;

	return {
		achievementPercentage,
		overallScore,
		rating,
	};
};

// API functions
export const performanceApi = {
	/**
	 * Get all performances for the current user
	 * @param accessToken JWT access token
	 * @returns Promise with array of performances
	 */
	getAll: async (accessToken: string): Promise<PerformanceResponse[]> => {
		const response = await fetch(`${API_BASE_URL}/performance`, {
			headers: createAuthHeaders(accessToken),
		}).then(handleApiError);

		return response.json();
	},

	/**
	 * Get performance by ID
	 * @param accessToken JWT access token
	 * @param id Performance ID
	 * @returns Promise with performance details
	 */
	getById: async (
		accessToken: string,
		id: number
	): Promise<PerformanceResponse> => {
		const response = await fetch(`${API_BASE_URL}/performance/${id}`, {
			headers: createAuthHeaders(accessToken),
		}).then(handleApiError);

		return response.json();
	},

	/**
	 * Create a new performance record
	 * @param accessToken JWT access token
	 * @param data Performance data
	 * @returns Promise with created performance
	 */
	create: async (
		accessToken: string,
		data: PerformanceRequest
	): Promise<PerformanceResponse> => {
		const response = await fetch(`${API_BASE_URL}/performance`, {
			method: 'POST',
			headers: createAuthHeaders(accessToken),
			body: JSON.stringify(data),
		}).then(handleApiError);

		return response.json();
	},

	/**
	 * Update a performance record
	 * @param accessToken JWT access token
	 * @param id Performance ID
	 * @param data Updated performance data
	 * @returns Promise with updated performance
	 */
	update: async (
		accessToken: string,
		id: number,
		data: Partial<PerformanceRequest>
	): Promise<PerformanceResponse> => {
		const response = await fetch(`${API_BASE_URL}/performance/${id}`, {
			method: 'PUT',
			headers: createAuthHeaders(accessToken),
			body: JSON.stringify(data),
		}).then(handleApiError);

		return response.json();
	},

	/**
	 * Get team performance (for managers)
	 * @param accessToken JWT access token
	 * @param startDate Start date
	 * @param endDate End date
	 * @returns Promise with array of team member performances
	 */
	getTeamPerformance: async (
		accessToken: string,
		startDate: string,
		endDate: string
	): Promise<PerformanceResponse[]> => {
		const response = await fetch(
			`${API_BASE_URL}/performance/team?startDate=${startDate}&endDate=${endDate}`,
			{
				headers: createAuthHeaders(accessToken),
			}
		).then(handleApiError);

		return response.json();
	},

	/**
	 * Add manager feedback
	 * @param accessToken JWT access token
	 * @param id Performance ID
	 * @param feedback Manager feedback
	 * @returns Promise with updated performance
	 */
	addManagerFeedback: async (
		accessToken: string,
		id: number,
		feedback: string
	): Promise<PerformanceResponse> => {
		const response = await fetch(`${API_BASE_URL}/performance/${id}/feedback`, {
			method: 'PUT',
			headers: createAuthHeaders(accessToken),
			body: JSON.stringify({ feedback }),
		}).then(handleApiError);

		return response.json();
	},
};

File: src/main/java/com/prime/prime_app/entities/Performance.java
package com.prime.prime_app.entities;

import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "performances", indexes = {
    @Index(name = "idx_performance_period", columnList = "periodStart,periodEnd"),
    @Index(name = "idx_agent_period", columnList = "agent_id,periodStart,periodEnd")
})
public class Performance {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "agent_id", nullable = false)
    private User agent;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "manager_id")
    private User manager;

    @Column(nullable = false)
    private LocalDate periodStart;

    @Column(nullable = false)
    private LocalDate periodEnd;

    @Column(nullable = false)
    @Min(value = 0, message = "New clients acquired cannot be negative")
    private Integer newClientsAcquired;

    @Column(nullable = false)
    @Min(value = 0, message = "Policies renewed cannot be negative")
    private Integer policiesRenewed;

    @Column(nullable = false)
    @Min(value = 0, message = "Total premium collected cannot be negative")
    private Double totalPremiumCollected;

    @Column(nullable = false)
    @Min(value = 0, message = "Sales target cannot be negative")
    private Double salesTarget;

    @Column(nullable = false)
    @Min(value = 0, message = "Sales achieved cannot be negative")
    private Double salesAchieved;

    @Column(nullable = false)
    private Double achievementPercentage;

    @Min(value = 0, message = "Client retention rate cannot be negative")
    @Max(value = 100, message = "Client retention rate cannot exceed 100")
    private Integer clientRetentionRate;

    @Min(value = 0, message = "Customer satisfaction score cannot be negative")
    @Max(value = 100, message = "Customer satisfaction score cannot exceed 100")
    private Integer customerSatisfactionScore;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PerformanceRating rating;

    @Column(length = 1000)
    private String managerFeedback;

    @Column(nullable = false)
    private Integer attendanceScore;

    @Column(nullable = false)
    private Integer qualityScore;

    @Column(nullable = false)
    private Double overallScore;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        calculateMetrics();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
        calculateMetrics();
        validatePeriod();
    }

    private void calculateMetrics() {
        // Calculate achievement percentage
        if (salesTarget != null && salesTarget > 0) {
            achievementPercentage = (salesAchieved / salesTarget) * 100;
        }

        // Calculate overall score (weighted average)
        double salesWeight = 0.4;
        double retentionWeight = 0.2;
        double satisfactionWeight = 0.2;
        double attendanceWeight = 0.1;
        double qualityWeight = 0.1;

        double salesScore = achievementPercentage != null ? achievementPercentage : 0;
        double retentionScore = clientRetentionRate != null ? clientRetentionRate : 0;
        double satisfactionScore = customerSatisfactionScore != null ? customerSatisfactionScore : 0;
        double attendanceScoreValue = attendanceScore != null ? attendanceScore : 0;
        double qualityScoreValue = qualityScore != null ? qualityScore : 0;

        overallScore = (salesScore * salesWeight) +
                      (retentionScore * retentionWeight) +
                      (satisfactionScore * satisfactionWeight) +
                      (attendanceScoreValue * attendanceWeight) +
                      (qualityScoreValue * qualityWeight);

        // Determine rating based on overall score
        rating = calculateRating(overallScore);
    }

    private PerformanceRating calculateRating(double score) {
        if (score >= 90) return PerformanceRating.OUTSTANDING;
        if (score >= 80) return PerformanceRating.EXCEEDS_EXPECTATIONS;
        if (score >= 70) return PerformanceRating.MEETS_EXPECTATIONS;
        if (score >= 60) return PerformanceRating.NEEDS_IMPROVEMENT;
        return PerformanceRating.UNSATISFACTORY;
    }

    private void validatePeriod() {
        if (periodStart != null && periodEnd != null && periodEnd.isBefore(periodStart)) {
            throw new IllegalStateException("Period end date cannot be before period start date");
        }
    }

    public enum PerformanceRating {
        OUTSTANDING,
        EXCEEDS_EXPECTATIONS,
        MEETS_EXPECTATIONS,
        NEEDS_IMPROVEMENT,
        UNSATISFACTORY
    }
}